### PR TITLE
feat(multimodal): add bottleneck fusion strategy

### LINKED ISSUE #253: MMT-104: Add Bottleneck Fusion Variants After MVP Stabilization
Актуализировано по текущему коду: 2026-06-15.

## Текущее состояние

В прототипе есть bottleneck-семейство:

- `ordinary_bottleneck`;
- `raw_residual_bottleneck`;
- `context_only_residual_bottleneck`;
- `raw_conditioned_context_bottleneck`.

По приложенным результатам bottleneck-семейство выглядит наиболее перспективным среди fusion-family aggregations: средний delta к single median около `+0.057`, медианный около `+0.039`, top family на 3 из 6 датасетов. Внутри bottleneck-вариантов лучший метод зависит от датасета: 
1. `context_only_residual_bottleneck` лучший на Crop, StarLightCurves и UWaveGestureLibraryAll
2.  `ordinary_bottleneck` лучший на FordA и NonInvasiveFetalECGThorax1, 
3. `raw_residual_bottleneck` лучший на ElectricDevices
4.  `raw_conditioned_context_bottleneck` в этих CSV ни разу не top-1 среди bottleneck-вариантов.

Риск - bottleneck сложнее, имеет больше параметров, attention/latent pooling могут переобучаться на малых датасетах и труднее диагностируются.

## Что сделать

После MVP из #252 добавить bottleneck-варианты как отдельный расширенный режим.

Первая очередь:

- `ordinary_bottleneck`;
- `raw_residual_bottleneck`;
- `context_only_residual_bottleneck`.

`raw_conditioned_context_bottleneck` добавить только если будет отдельное обоснование или ablation, потому что в приложенных результатах он не был лидером.

Для bottleneck-моделей обязательно сохранить diagnostics:

- attention summary;
- pooling mode;
- `num_latents`, `num_heads`, `num_layers`;
- alpha/residual strength, если есть residual path;
- parameter count и train duration.

## Что уже не надо делать повторно

1. Не переписывать encoder registry и preprocessing. Использовать #251 и #252.

2. Не включать все bottleneck-варианты в default preset до появления smoke-тестов и сравнения с MiniRocket.

## Критерии готовности

1. Bottleneck models собираются тем же публичным config path, что и MVP fusion models.

2. Forward-pass и короткое обучение на synthetic dataset проходят на CPU.

3. Diagnostics позволяют понять, не схлопнулся ли residual/gating/attention путь.

## Проверки

1. Unit-тесты на shape, missing modalities, invalid `num_heads`, invalid pooling.

2. Smoke-тест обучения на маленьком synthetic dataset с `epochs=1..2`.

3, Benchmark smoke через новый adapter из следующей задачи.

## Зависимости

Зависит от #252.

### PR DESCRIPTION
**Pull request by Romankkl03**
## Summary
- Add extended bottleneck fusion methods (`ordinary`, `raw_residual`, `context_only`) to the FUTURE registry with attention/pooling/residual diagnostics.
- Introduce a bundle-aware `FutureClassifierTrainer` with aligned multimodal batching, configurable optimizer/criterion, and device-aware timing (`DeviceTimer`).
- Wire a thin benchmark adapter (`future_fusion_classifier`) and replace the classification adapter `if`-chain with a class-based registry.

## Test plan
- [x] Unit tests for bottleneck shapes, invalid pooling/heads, residual diagnostics, missing modalities
- [x] Trainer CPU tests (fit/predict, early stopping, custom optimizer, history/checkpoints)
- [x] Optional CUDA `DeviceTimer` test (`skipif` when no GPU)
- [x] Benchmark adapter smoke + unknown-adapter registry error

Closes #253

### CODE REVIEW

File: benchmark/industrial/models/classification.py [hunk 1]
```diff
@@ -138,33 +138,251 @@ def predict(self, features: np.ndarray) -> np.ndarray:
...
+            artifacts['d_model'] = self.trainer_.model.d_model
+            artifacts['num_parameters'] = {
+                'total': int(
+                    sum(
+                        parameter.numel()
+                        for parameter in self.trainer_.model.parameters()
+                    )
+                ),
+            }
+        return artifacts
```
Comment (Lopa10ko) [inline benchmark/industrial/models/classification.py:340]: the benchmark suite already calls `export_artifacts()` and the incremental writer already only persists `kernel_diagnostics` / `kernel_selection` / `summary` - **neither path is changed in this PR**. So these new FUTURE keys are silently discarded at runtime!!

either 
(a) extend the writer to persist any non-empty `model_artifacts` (e.g. `model_artifacts.json`)
(b) stop advertising FUTURE history via `export_artifacts` until the writer supports it, and add a suite/adapter test that fails if FUTURE keys never hit disk

File: benchmark/industrial/models/classification.py [hunk 2]
```diff
@@ -138,33 +138,251 @@ def predict(self, features: np.ndarray) -> np.ndarray:
...
+                'FutureFusionClassifierAdapter must be fitted before prediction.'
+            )
+        test_bundle = self.preparer_.transform(features)
+        predictions = (
+            self.trainer_.predict(test_bundle.without_target()).detach().cpu().numpy()
+        )
+        return np.asarray(
+            [self.inverse_label_mapping_[int(index)] for index in predictions],
+            dtype=object,
+        )
```
Comment (Lopa10ko) [inline benchmark/industrial/models/classification.py:307]: always remap to contiguous `0..C-1` inside this new adapter branch (do not rely on raw label ids as class indices)

decode with an explicit OOV error, add a decoding unit test

File: benchmark/industrial/models/classification.py [hunk 3]
```diff
@@ -138,33 +138,251 @@ def predict(self, features: np.ndarray) -> np.ndarray:
...
-            optional=True,
-            params=dict(spec.params),
+        training_config = FutureTrainingConfig(
+            epochs=int(training_kwargs.pop('epochs', 2)),
+            batch_size=int(training_kwargs.pop('batch_size', 32)),
+            learning_rate=float(training_kwargs.pop('learning_rate', 1e-3)),
+            weight_decay=float(training_kwargs.pop('weight_decay', 0.0)),
+            early_stopping_patience=training_kwargs.pop('early_stopping_patience', None),
+            device=training_kwargs.pop('device', 'cpu'),
+            seed=training_kwargs.pop('seed', 42),
```
Comment (Lopa10ko) [inline benchmark/industrial/models/classification.py]: why use defaults here? `FutureTrainingConfig` already has defaults on init and they are different!!

File: benchmark/industrial/models/classification.py [hunk 4]
```diff
@@ -138,33 +138,251 @@ def predict(self, features: np.ndarray) -> np.ndarray:
...
+        if classifier_kwargs:
+            raise BenchmarkClassificationError(
+                f'Unsupported FUTURE adapter params: {sorted(classifier_kwargs)}'
+            )
+
+        self.preparer_ = MultimodalDatasetPreparer(
+            config=build_preparation_config(**preparation_kwargs)
+            if preparation_kwargs
+            else build_preparation_config()
+        )
```
Comment (Lopa10ko) [inline benchmark/industrial/models/classification.py:205]: new code pops `modalities` for the classifier, then builds an unrestricted default preparation config when `preparation` is omitted

so `modalities=['raw']` still pays for the full default prep stack (raw/stats/gaf/stft) unless the caller also passes `preparation.transformation_config`

when `preparation` is omitted, derive `transformation_config` from `modalities` (default `['raw']` if `modalities is None`)

File: benchmark/industrial/models/classification.py [hunk 5]
```diff
@@ -61,7 +61,7 @@ def predict(self, features: np.ndarray) -> np.ndarray:
 class OptionalExternalClassifier:
     dependency_name: str
     name: str
-    tags: tuple[str, ...] = ('baseline', 'classification', 'external')
+    tags: tuple[str, ...] = ('industrial', 'classification', 'external')
```
Comment (Lopa10ko) [inline benchmark/industrial/models/classification.py]: unrelated tag change, why so?

File: fedot_ind/core/models/future/mapping.py [hunk 6]
```diff
@@ -17,24 +23,50 @@ class FusionRegistryEntry:
     """Registry descriptor for one FUTURE fusion strategy."""
 
     fusion_class: type[BaseFusionStrategy]
-    inputs_param_name: str
+    inputs_param: FusionInputsParam
```
Comment (Lopa10ko) [inline fedot_ind/core/models/future/mapping.py:26]: check if this change (`inputs_param_name: str` to `inputs_param: FusionInputsParam`) lives only in FUTURE registries

File: fedot_ind/core/models/future/trainer.py [hunk 7]
```diff
@@ -0,0 +1,391 @@
...
+    def evaluate_diagnostics(self, bundle: MultimodalDataBundle) -> FusionAuxOutput:
+        """Run a diagnostic forward pass with fusion aux payload."""
+
+        self._ensure_built(bundle)
+        self.model.eval()
+        device_bundle = bundle.to(device=self.device)
+        aux = self.model(device_bundle, return_aux=True)
+        if not isinstance(aux, FusionAuxOutput):
+            raise RuntimeError("Expected FusionAuxOutput when return_aux=True.")
+        return aux
```
Comment (Lopa10ko) [inline fedot_ind/core/models/future/trainer.py]: batch like `predict`, and/or default aux embeddings off for this helper

File: fedot_ind/core/models/future/trainer.py [hunk 8]
```diff
@@ -0,0 +1,391 @@
...
+                f"got dtype={bundle.target.dtype}."
+            )
+
+    @staticmethod
+    def _set_seed(seed: int | None) -> None:
+        if seed is None:
+            return
+        torch.manual_seed(seed)
+        if torch.cuda.is_available():
+            torch.cuda.manual_seed_all(seed)
```
Comment (Lopa10ko) [inline fedot_ind/core/models/future/trainer.py]: it should be a utility inside `fedot_ind` or maybe a similar method already exists

File: fedot_ind/core/models/future/trainer.py [hunk 9]
```diff
@@ -0,0 +1,391 @@
...
+    def load_checkpoint(
+        cls,
+        path: str | Path,
+        *,
+        device: Any = "cpu",
+        config: FutureTrainingConfig | None = None,
+    ) -> "FutureClassifierTrainer":
+        """Restore a trainer and rebuilt model from ``save_checkpoint`` output."""
+
+        payload = torch.load(Path(path), map_location="cpu", weights_only=False)
```
Comment (Lopa10ko) [inline fedot_ind/core/models/future/trainer.py]: prefer `weights_only=True` if possible for state dict + JSON/config

Comment (Lopa10ko) [inline fedot_ind/core/models/future/trainer.py]: also, shouldn't the payload extraction be mapped to `device` instead a hard-set `cpu`?

**Review by Lopa10ko — APPROVED** [review]

File: fedot_ind/core/models/future/trainer.py (hunk 9 above)
Comment (Romankkl03) [reply fedot_ind/core/models/future/trainer.py]: Added device resolving in fc996e0

Comment (Romankkl03) [reply fedot_ind/core/models/future/trainer.py]: Done in 0a38e98. Now, weights and meta are saved separately.

File: fedot_ind/core/models/future/trainer.py (hunk 8 above)
Comment (Romankkl03) [reply fedot_ind/core/models/future/trainer.py]: There were not utility like this, so moved to torch_backend/io.py in 804b184

File: fedot_ind/core/models/future/trainer.py (hunk 7 above)
Comment (Romankkl03) [reply fedot_ind/core/models/future/trainer.py]: refactored in d21d79d

File: fedot_ind/core/models/future/mapping.py (hunk 6 above)
Comment (Romankkl03) [reply fedot_ind/core/models/future/mapping.py:26]: Checked: inputs_param / FusionInputsParam are only used under fedot_ind/core/models/future/ (+ unit test). No callers outside FUTURE depended on inputs_param_name.

File: benchmark/industrial/models/classification.py (hunk 5 above)
Comment (Romankkl03) [reply benchmark/industrial/models/classification.py]: I’ve reverted the class default to baseline (aligned with OptionalExternalRegressor) and moved the industrial tags into _ADAPTER_EXTRA_KWARGS in 95a4f72

File: benchmark/industrial/models/classification.py (hunk 4 above)
Comment (Romankkl03) [reply benchmark/industrial/models/classification.py:205]: Refactored and added tests in d812d7c

File: benchmark/industrial/models/classification.py (hunk 3 above)
Comment (Romankkl03) [reply benchmark/industrial/models/classification.py]: In 5a6d7e1 deleted unnecessary defaults assigning

File: benchmark/industrial/models/classification.py (hunk 2 above)
Comment (Romankkl03) [reply benchmark/industrial/models/classification.py:307]: In a1d73eb added class ContiguousLabelEncoder in fedot_ind/core/architecture/preprocessing/label_mapping.py with LabelMappingError(ValueError) and added unit tests

File: benchmark/industrial/models/classification.py (hunk 1 above)
Comment (Romankkl03) [reply benchmark/industrial/models/classification.py:340]: Extend the writer to persist any non-empty model_artifacts in ca5c6ce
